halo2_proofs/src/plonk.rs

```diff
@@ -44,17 +44,15 @@ use std::io;
 pub fn vk_read<C: SerdeCurveAffine, R: io::Read, ConcreteCircuit: Circuit<C::Scalar>>(
     reader: &mut R,
     format: SerdeFormat,
+    k: u32,
+    circuit: &ConcreteCircuit,
     compress_selectors: bool,
-    #[cfg(feature = "circuit-params")] params: ConcreteCircuit::Params,
 ) -> io::Result<VerifyingKey<C>>
 where
     C::Scalar: SerdePrimeField + FromUniformBytes<64>,
 {
-    let (_, cs, _) = compile_circuit_cs::<_, ConcreteCircuit>(
-        compress_selectors,
-        #[cfg(feature = "circuit-params")]
-        params,
-    );
+    let (_, _, cs) = compile_circuit(k, circuit, compress_selectors)
```

Comment:
This is why I tried to implement `compile_circuit_cs` originally.  So that we could do `vk_read` without having to call synthesize on the circuit :(
But unfortunately it's not possible; the cs needs the conversion of selector expressions; and that requires selector assignments, which we obtain by calling synthesize in `compile_circuit`.

As a result the methods `vk_read` and `pk_read` do a lot of work.

Nevertheless the frontend-backend split API methods for reading vk and pk don't suffer from this: https://github.com/privacy-scaling-explorations/halo2/blob/9eedeb5d6eb9a119d34307697e670b0bb043a5a5/halo2_backend/src/plonk.rs#L99
So I think this is acceptable :)